compose-builder/tui-generator.sh

```diff
@@ -0,0 +1,293 @@
+
+# Individual Compose Descriptions that may not fall into EdgeX-Foundry Architecture Components
+# Or may be an initial "out-of-the-box" configuration
+
+MAKE=/usr/bin/make
+
+SELECTED_DEVSERVICES=()
+SELECTED_APPSERVICES=()
+SELECTED_BUS=()
+SELECTED_OTHERS=()
+
+## Additional Options
+additionalOpts=(
+    "mqtt-broker"
+    "modbus-sim"
+)
+
+## General Options Description
+declare -A additionalOptsDesc=(
+    [modbus-sim]="Include ModBus Simulator"
+    [mqtt-broker]="Include MQTT Broker"
+)
+
+
+## Available App Services in EdgeX-Foundry
+appServices=(
+    "asc-http"
+    "asc-mqtt"
+    "asc-sample"
+    "as-llrp"
+    "asc-ex-mqtt"
+)
+
+## App Service Descriptions
+declare -A appServiceDesc=(
+    [asc-http]="Include HTTP Export App Service"
+    [asc-mqtt]="Include MQTT Export App Service"
+    [asc-sample]="Include Sample App Service"
+    [as-llrp]="Include RFID LLRP Inventory"
+    [asc-ex-mqtt]="Include External MQTT Trigger App Service"
+)
+
+
+## Available Device Services in EdgeX-Foundry
+deviceServices=(
+    "ds-bacnet"
+    "ds-camera"
+    "ds-grove"
+    "ds-modbus"
+    "ds-mqtt"
+    "ds-rest"
+    "ds-snmp"
+    "ds-virtual"
+    "ds-coap"
+    "ds-gpio"
+    "ds-llrp"
+)
+
+## Device Service Descriptions
+declare -A deviceServiceDesc=(
+    [ds-bacnet]="Include BACnet Device Service"
+    [ds-camera]="Include Camera Device Service"
+    [ds-grove]="Include Grove Device Service (valid only: arm64)"
+    [ds-modbus]="Include ModBus Device Service"
+    [ds-mqtt]="Include MQTT Device Service"
+    [ds-rest]="Include REST Device Service"
+    [ds-snmp]="Include SNMP Device Service"
+    [ds-virtual]="Include Virtual Device Service"
+    [ds-coap]="Include CoAP Device Service"
+    [ds-gpio]="Include GPIO Device Service"
+    [ds-llrp]="Include RFID LLRP Device Service"
+
+)
+
+
+## Available Message Bus Services in EdgeX-Foundry
+messageBus=(
+    "mqtt-bus"
+    "zmq-bus"
+)
+
+
+## Message Bus Descriptions
+declare -A msgBusDesc=(
+    [mqtt-bus]="Configure MQTT Message Bus"
+    [zmq-bus]="Configure ZMQ Message Bus"
+)
+
+
+####################################################################
+#    Additional Services Options Display Function                  #
+####################################################################
+function additionalServiceOption() {
+    message="Some Additional Services are available that can also be included. \n
+            Press <SPACEBAR> to Select \n
+            Press <ENTER> to Skip \n
+            "
+
+    # Generate a Specific form of Array required by whiptail checklist
+    # FORMAT: "<INDEX> <DESCRIPTION> <OFF>"
+    arglist=()
+    for index in "${additionalOpts[@]}";
+    do
+        arglist+=("$index")
+        arglist+=("${additionalOptsDesc[$index]}")
+        arglist+=("OFF") # Nothing is selected by-default
+    done
+    SELECTED_OTHERS+=$(/usr/bin/whiptail --title "App Services" \
```

Comment:
Don't assume where whiptail is install. This fails for me. Just assume it is in the user's path. You may what to add a check that it is available and fail out if not.
```suggestion
    SELECTED_OTHERS+=$(whiptail --title "App Services" \
```

Reply:
This is resolved now by initially checking whether `whiptail` exists, if not exit the script. If it does exist the path to the binary will be stored in a variable and then resolved at all places where `whiptail` was initially used.